source/byte_buf.c

```diff
@@ -242,32 +259,102 @@ static const uint8_t s_tolower_table[256] = {
     220, 221, 222, 223, 224, 225, 226, 227, 228, 229, 230, 231, 232, 233, 234, 235, 236, 237, 238, 239, 240, 241,
     242, 243, 244, 245, 246, 247, 248, 249, 250, 251, 252, 253, 254, 255};
 
-bool aws_byte_cursor_eq_case_insensitive(const struct aws_byte_cursor *a, const struct aws_byte_cursor *b) {
-    if (!a || !b) {
-        return a == b;
+bool aws_array_eq_ignore_case(const void *array_a, size_t len_a, const void *array_b, size_t len_b) {
+    if (len_a != len_b) {
+        return false;
     }
 
-    if (a->len != b->len) {
-        return false;
+    /* If either is null: true if both are null or the size is 0 */
+    if (!array_a || !array_b) {
+        return (array_a == array_b) || (len_a == 0);
     }
 
-    for (size_t i = 0; i < a->len; ++i) {
-        if (s_tolower_table[a->ptr[i]] != s_tolower_table[b->ptr[i]]) {
+    const uint8_t *bytes_a = array_a;
+    const uint8_t *bytes_b = array_b;
+    for (size_t i = 0; i < len_a; ++i) {
+        if (s_tolower_table[bytes_a[i]] != s_tolower_table[bytes_b[i]]) {
             return false;
         }
     }
 
     return true;
 }
 
-uint64_t aws_hash_byte_cursor_ptr_case_insensitive(const void *item) {
+bool aws_array_eq(const void *array_a, size_t len_a, const void *array_b, size_t len_b) {
+    if (len_a != len_b) {
+        return false;
+    }
+
+    /* If either is null: true if both are null or the length is 0 */
+    if (!array_a || !array_b) {
+        return (array_a == array_b) || (len_a == 0);
+    }
+
+    return !memcmp(array_a, array_b, len_a);
+}
+
+bool aws_array_eq_c_str_ignore_case(const void *array, size_t array_len, const char *c_str) {
+    const uint8_t *array_bytes = array;
+    const uint8_t *str_bytes = (const uint8_t *)c_str;
+
+    if (!str_bytes) {
+        return array_len == 0;
+    }
+
+    if (!array_bytes && (array_len != 0)) {
+        return false;
+        /* If array_len is 0 we still must check that c_str is null terminated */
+    }
+
+    for (size_t i = 0; i < array_len; ++i) {
+        uint8_t s = str_bytes[i];
+        if (s == '\0') {
+            return false;
+        }
+
+        if (s_tolower_table[array_bytes[i]] != s_tolower_table[s]) {
+            return false;
+        }
+    }
+
+    return str_bytes[array_len] == '\0';
+}
+
+bool aws_array_eq_c_str(const void *array, size_t array_len, const char *c_str) {
+    const uint8_t *array_bytes = array;
```

Comment:
Why doesn't this strlen() and delegate to ```aws_array_eq()```?

Reply:
Calling strlen() means traversing the c-str an additional time.

This function is very copy/pasty similar to the other one, but checks for null-terminators in the c_str while doing the comparison, so there's only one traversal.

I could benchmark the two implementations to see if there's measurable difference.

Thoughts?

Reply:
Oh, yeah, of course. That makes sense, missed that one :)

Reply:
Did some benchmarks and it is worth avoiding the strlen.
- 2X speedup when avoiding it in `aws_array_eq_cstr()` (uses memcmp under the hood)
- 1.5X speedup when avoiding it in `aws_array_eq_c_str_ignore_case()` (uses my crazy table under the hood)